### PR TITLE
i18n: add locale variants to the server side

### PR DESCRIPTION
**Pull request by southp**
## Summary
This PR is part of the on-going work of splitting #23025 into smaller, shippable pieces and it's based on #23152 . In this PR, we introduce locale variants to the server side.

### CODE REVIEW

**Comment by southp** [comment]
@ockham I'm cc-ing you here since you are the go-to person for everything SSR to me :)

For Team Global's hack week project, we are implementing the foundation of introducing "locale variants". A locale variant is basically a variation of a locale, which is mostly the same language but with some syntactical changes. The first variant we are going to introduce, formal Germany ( `de_formal` ) is a very good example on this aspect. More details can be found in #23025 and p2-pxLjZ-4wi.

I don't know much about our SSR setup, so basically I'd like to ask are:

1. How do I test my code locally? I'm worrying that I might have broken something without knowing.
1. Is the code we added here sufficient for covering locale changes from server side?

Thanks 🙇

File: server/render/index.js [hunk 1]
```diff
@@ -152,7 +152,10 @@ export function serverRender( req, res ) {
 			stateCache.set( cacheKey, serverState );
 		}
 
-		context.lang = getCurrentLocaleSlug( context.store.getState() ) || context.lang;
+		context.lang =
+			getCurrentLocaleVariant( context.store.getState() ) ||
```
Comment (ramonjd) [inline server/render/index.js]: I think I missed this last time, but, unless we're using the locale variant explicitly, I'm not convinced we need to set `context.lang` to the variant at all here.

`context.lang` is assigned and used in a few places, one example is setting the document title. With user bootstrapping on, this is the side effect of the above expression:

screen shot 2018-03-13 at 12 45 33 pm

Ideally it should be `de`.

I've tested with just the `localeSlug`, and there appears to be no side effects related to removing the locale variant check:

```
		context.lang = getCurrentLocaleSlug( context.store.getState() ) || context.lang;
```

There is one other issue however, and that is the value we assign to `context.i18nLocaleScript`. As far as I can tell, for user-bootstrapped pages we use the values in this script to [populate the locale strings](https://github.com/Automattic/wp-calypso/blob/151b0105a5734b6a60def82d5079370fa76afe28/client/boot/common.js#L144).

So, even with a locale variant set, the result is the following:

screen shot 2018-03-13 at 12 59 55 pm

I think we could get around this by [checking for the variant earlier at this line](https://github.com/Automattic/wp-calypso/blob/151b0105a5734b6a60def82d5079370fa76afe28/server/render/index.js#L116):

```
	if ( ! isDefaultLocale( context.lang ) ) {
		const langFileName = getCurrentLocaleVariant( context.store.getState() ) || context.lang;
		context.i18nLocaleScript = '//widgets.wp.com/languages/calypso/' + langFileName + '.js';
	}
```

Comment (southp) [reply server/render/index.js]: Nice catch! I've updated the code in f9a5904.

**Comment by seear** [comment]
> I don't know much about our SSR setup, so basically I'd like to ask are:
> 1. How do I test my code locally? I'm worrying that I might have broken something without knowing.
> 2. Is the code we added here sufficient for covering locale changes from server side?

@southp Currently, as far as I'm aware, Server-side-rendering (as in serving pre-rendered, cached content from the server) is not enabled for any locales other than the default. We can see that from this check here https://github.com/Automattic/wp-calypso/pull/23154/files#diff-d92d930a21711a960b300aeb231d9056R126.

So, this PR should not have any effect on SSR'd content.

As I understand it, this change will affect only logged-in users with accounts set to non-default locale, and sets the name of the language script in the served page, for example:
`<script src="//widgets.wp.com/languages/calypso/fr.js">`
Is that correct?

For general tips on testing server-side changes:
* Remember to restart the build after every change, because only client-side changes force auto-rebuild :(
* It can be useful to turn on debug before starting the build, using:
`export DEBUG="calypso:server-render"`

**Comment by ramonjd** [comment]
> As I understand it, this change will affect only logged-in users with accounts set to non-default locale, and sets the name of the language script in the served page ... Is that correct?

Thanks for confirming @seear! 

Yes, we're only interested in whether `localeVariant` is set on the user object to load the language script. We implemented and merged this in [another PR](https://github.com/Automattic/wp-calypso/pull/23162/files#diff-b7d5107c83e8826b53a71a100d797652). If `localeVariant` isn't set it will work as it always has.

**Comment by ramonjd** [comment]
Merging for @southp

**Pull request merged**